### PR TITLE
Story/ccls 1734 copy case

### PR DESCRIPTION
**Pull request by porritta**

### CODE REVIEW

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 1]
```diff
@@ -210,5 +254,135 @@ public Mono<UserDetails> getUsers(final Integer providerId) {
 
   }
 
+  /**
+   * Retrieves proceeding detail.
+   *
+   * @return A Mono containing the ProceedingDetail or an error handler if an error occurs.
+   */
+  public Mono<ProceedingDetail> getProceeding(String proceedingCode) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:262]: missing @param in javadoc
final on params

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 2]
```diff
@@ -210,5 +254,135 @@ public Mono<UserDetails> getUsers(final Integer providerId) {
...
+        .onErrorResume(e -> ebsApiClientErrorHandler.handleProceedingError(proceedingCode, e));
+  }
+
+  /**
+   * Retrieves scope limitation details which match the provided example ScopeLimitationDetail.
+   *
+   * @return A Mono containing the ScopeLimitationDetails or an error handler if an error occurs.
+   */
+  public Mono<ScopeLimitationDetails> getScopeLimitations(
+      ScopeLimitationDetail scopeLimitationDetail) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:277]: missing @param in javadoc
final on params

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 3]
```diff
@@ -58,36 +65,41 @@ public Mono<ProviderDetail> getProvider(Integer providerId) {
...
    *
    * @param type The type of the common lookup values. Can be null.
    * @param code The code of the common lookup values. Can be null.
+   * @param description The description of the common lookup values. Can be null.
    * @param sort The sort criteria for the common lookup values. Can be null.
    * @return A Mono containing the CommonLookupDetail or an error handler if an error occurs.
    */
-  public Mono<CommonLookupDetail> getCommonValues(String type, String code, String sort) {
+  public Mono<CommonLookupDetail> getCommonValues(String type, String code,
+      String description, String sort) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:77]: finals

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 4]
```diff
@@ -58,36 +65,41 @@ public Mono<ProviderDetail> getProvider(Integer providerId) {
...
    *
    * @param type The type of the common lookup values. Can be null.
+   * @param code The code of the common lookup values. Can be null.
+   * @param description The description for the common lookup values. Can be null.
    * @return A Mono containing the CommonLookupDetail or an error handler if an error occurs.
    */
-  public Mono<CommonLookupDetail> getCommonValues(String type) {
-    return this.getCommonValues(type, null, null);
+  public Mono<CommonLookupDetail> getCommonValues(String type, String code, String description) {
+    return this.getCommonValues(type, code, description, null);
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:102]: finals

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 5]
```diff
@@ -101,6 +113,17 @@ public Mono<CommonLookupDetail> getCommonValues(String type, String code) {
     return this.getCommonValues(type, code, null);
   }
 
+  /**
+   * Retrieves common lookup values based on the supplied type.
+   *
+   * @param type The type of the common lookup values. Can be null.
+   * @return A Mono containing the CommonLookupDetail or an error handler if an error occurs.
+   */
+  public Mono<CommonLookupDetail> getCommonValues(String type) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:122]: final

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 6]
```diff
@@ -190,6 +213,27 @@ public Mono<CommonLookupDetail> getCountries() {
             .onErrorResume(ebsApiClientErrorHandler::handleCountryLookupError);
   }
 
+  /**
+   * Retrieves prior authority types by code and valueRequired flag.
+   *
+   * @return A Mono containing the PriorAuthorityTypeDetails or an error handler if an error occurs.
+   */
+  public Mono<PriorAuthorityTypeDetails> getPriorAuthorityTypes(String code,
+      Boolean valueRequired) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:222]: finals

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 7]
```diff
@@ -210,5 +254,135 @@ public Mono<UserDetails> getUsers(final Integer providerId) {
...
+  /**
+   * Retrieves outcome result lookup detail based on the provided proceedingCode and
+   * outcomeResult values.
+   *
+   * @param proceedingCode - the proceeding code.
+   * @param outcomeResult - the outcome result value.
+   * @return A Mono containing the ProceedingDetail or an error handler if an error occurs.
+   */
+  public Mono<OutcomeResultLookupDetail> getOutcomeResults(String proceedingCode,
+      String outcomeResult) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:325]: final on params

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 8]
```diff
@@ -210,5 +254,135 @@ public Mono<UserDetails> getUsers(final Integer providerId) {
...
+  /**
+   * Retrieves stage end lookup detail based on the provided proceedingCode and
+   * stageEnd values.
+   *
+   * @param proceedingCode - the proceeding code.
+   * @param stageEnd - the stage end value.
+   * @return A Mono containing the StageEndLookupDetail or an error handler if an error occurs.
+   */
+  public Mono<StageEndLookupDetail> getStageEnds(String proceedingCode,
+      String stageEnd) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:349]: final on params

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 9]
```diff
@@ -210,5 +254,135 @@ public Mono<UserDetails> getUsers(final Integer providerId) {
...
+  /**
+   * Retrieves award type lookup detail based on the provided code and
+   * award type values.
+   *
+   * @param code - the award type code.
+   * @param awardType - the award type value.
+   * @return A Mono containing the AwardTypeLookupDetail or an error handler if an error occurs.
+   */
+  public Mono<AwardTypeLookupDetail> getAwardTypes(String code,
+      String awardType) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:373]: final on params

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 10]
```diff
@@ -73,11 +122,27 @@ public Mono<CaseDetails> getCases(CopyCaseSearchCriteria copyCaseSearchCriteria,
...
 
+  /**
+   * Retrieve the full details of a Case.
+   *
+   * @param caseReferenceNumber The reference of the case to be retrieved.
+   * @param loginId             The login identifier for the user.
+   * @param userType            Type of the user (e.g., admin, user).
+   * @return A Mono wrapping the CaseDetails.
+   */
+  public Mono<ApplicationDetail> getCase(String caseReferenceNumber, String loginId,
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:133]: finals

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 11]
```diff
@@ -333,5 +398,451 @@ public void patchProviderDetails(
...
 
+  /**
+   * Before a CaseDetail can be mapped to a CAAB ApplicationDetail further lookup
+   * data and calculations need to be performed. This method builds a wrapper object to
+   * hold all the required data for the mapping.
+   *
+   * @param soaCase - the SOA CaseDetail.
+   * @return a Mono containing an ApplicationMappingContext for the CaseDetail.
+   */
+  protected Mono<ApplicationMappingContext> buildApplicationMappingContext(CaseDetail soaCase) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:409]: add final to param

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 12]
```diff
@@ -333,5 +398,451 @@ public void patchProviderDetails(
...
+    });
+  }
+
+  protected Mono<ProceedingMappingContext> buildProceedingMappingContext(
+      ProceedingDetail soaProceeding,
+      CaseDetail soaCase,
+      Map<String, CommonLookupValueDetail> matterTypeLookups,
+      Map<String, CommonLookupValueDetail> levelOfServiceLookups,
+      Map<String, CommonLookupValueDetail> clientInvolvementLookups,
+      Map<String, CommonLookupValueDetail> scopeLimitationLookups) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:580]: add finals to params

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 13]
```diff
@@ -333,5 +398,451 @@ public void patchProviderDetails(
...
+  /**
+   * Build a mapping context to hold a SOA PriorAuthority and associated
+   * lookup data.
+   *
+   * @param soaPriorAuthority - the PriorAuthority to map.
+   * @param priorAuthorityTypes - all PriorAuthorityTypes.
+   * @return a PriorAuthorityMappingContext containing all data to support mapping
+   *    to a CAAB PriorAuthority.
+   */
+  protected PriorAuthorityMappingContext buildPriorAuthorityMappingContext(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:678]: add finals to params

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 14]
```diff
@@ -333,5 +398,451 @@ public void patchProviderDetails(
...
+          .soaPriorAuthority(soaPriorAuthority)
+          .priorAuthorityTypeLookup(priorAuthorityType)
+          .items(priorAuthorityDetails)
+          .build();
+  }
+
+  private CommonLookupValueDetail getPriorAuthLookup(
+      PriorAuthorityDetail priorAuthorityDetail,
+      PriorAuthorityAttribute priorAuthorityAttribute) {
+    String description;
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:719]: add finals to params

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 15]
```diff
@@ -333,5 +398,451 @@ public void patchProviderDetails(
...
+    } else {
+      description = priorAuthorityAttribute.getValue();
+    }
+
+    return new CommonLookupValueDetail()
+        .code(priorAuthorityAttribute.getValue())
+        .description(description);
+  }
+
+  protected CaseOutcomeMappingContext buildCaseOutcomeMappingContext(CaseDetail soaCase,
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:740]: add finals to params

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java [hunk 16]
```diff
@@ -333,5 +398,451 @@ public void patchProviderDetails(
...
+        .costAwards(costAwards)
+        .financialAwards(financialAwards)
+        .landAwards(landAwards)
+        .otherAssetAwards(otherAssetAwards)
+        .proceedingOutcomes(proceedingMappingContexts)
+        .build();
+  }
+
+  protected BigDecimal calculateProceedingCostLimitation(ProceedingDetail proceeding,
+      CaseDetail soaCase) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:785]: add finals to params

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 17]
```diff
@@ -258,7 +256,172 @@ public Mono<CommonLookupDetail> getDisabilities() {
...
+  public Mono<CommonLookupDetail> getScopeLimitations() {
+    return ebsApiClient.getCommonValues(COMMON_VALUE_SCOPE_LIMITATIONS);
+  }
+
+  /**
+   * Get a list of Scope Limitation Detail.
+   *
+   * @return ScopeLimitationDetails containing the scope limitation details.
+   */
+  public Mono<ScopeLimitationDetails> getScopeLimitationDetails(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:305]: add finals to params

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 18]
```diff
@@ -258,7 +256,172 @@ public Mono<CommonLookupDetail> getDisabilities() {
...
+  public Mono<CommonLookupDetail> getProceedingStatuses() {
+    return ebsApiClient.getCommonValues(COMMON_VALUE_PROCEEDING_STATUS);
+  }
+
+  /**
+   * Get a single Proceeding Status Lookup Value by code.
+   *
+   * @return CommonLookupValueDetail containing the status value.
+   */
+  public Mono<CommonLookupValueDetail> getProceedingStatus(String code) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:324]: add param to java doc

add finals to params

File: src/main/resources/application-local.yml
Comment (PhilDigitalJustice) [inline src/main/resources/application-local.yml:1]: Did you mean to push this file?

**Review by PhilDigitalJustice — COMMENTED** [review]
A few comments to resolve :)

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 19]
```diff
@@ -190,6 +214,27 @@ public Mono<CommonLookupDetail> getCountries() {
             .onErrorResume(ebsApiClientErrorHandler::handleCountryLookupError);
   }
 
+  /**
+   * Retrieves prior authority types by code and valueRequired flag.
+   *
+   * @return A Mono containing the PriorAuthorityTypeDetails or an error handler if an error occurs.
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:220]: missing param in java doc

File: src/main/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapper.java [hunk 20]
```diff
@@ -0,0 +1,645 @@
...
+  @Mapping(target = "ebsId", ignore = true)
+  @Mapping(target = "auditTrail", ignore = true)
+  PriorAuthority toPriorAuthority(PriorAuthorityMappingContext priorAuthorityMappingContext);
+
+  @Mapping(target = "code.id", source = "key.code")
+  @Mapping(target = "code.displayValue", source = "key.description")
+  @Mapping(target = "type", source = "key.dataType")
+  @Mapping(target = "lovLookUp", source = "key.lovCode")
+  @Mapping(target = "value", source = "value")
+  //  @Mapping(target = "mandatory", source = "mandatoryFlag")
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapper.java:318]: need to remove this?

File: src/main/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapper.java [hunk 21]
```diff
@@ -0,0 +1,645 @@
...
+
+  /**
+   * Complete the mapping for an OtherAssetAward.
+   *
+   * @param otherAssetAward - the mapping target
+   */
+  @AfterMapping
+  default void finaliseOtherAssetAward(@MappingTarget OtherAssetAward otherAssetAward) {
+    TimeRecovery timeRecovery = otherAssetAward.getTimeRecovery();
+    /* TODO: To be changed once api updated */
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapper.java:500]: Make sure we do this!

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 22]
```diff
@@ -0,0 +1,466 @@
...
+    return ebsApiClient.getCommonValues(COMMON_VALUE_GENDER);
+  }
+
+  /**
+   * Get a Gender Common Values.
+   *
+   * @param code The code of the common lookup values.
+   * @return CommonLookupValueDetail containing the common lookup value.
+   */
+  public Mono<CommonLookupValueDetail> getGender(String code) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:86]: Final

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 23]
```diff
@@ -0,0 +1,466 @@
...
+    return ebsApiClient.getCommonValues(COMMON_VALUE_CONTACT_TITLE);
+  }
+
+  /**
+   * Get a Contact Title Common Value.
+   *
+   * @param code The code of the common lookup values
+   * @return CommonLookupDetail containing the common lookup values.
+   */
+  public Mono<CommonLookupValueDetail> getContactTitle(String code) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:114]: final

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 24]
```diff
@@ -0,0 +1,466 @@
...
+    return ebsApiClient.getCommonValues(COMMON_VALUE_MARITAL_STATUS);
+  }
+
+  /**
+   * Get a Marital Status Common Value.
+   *
+   * @param code The code of the common lookup values
+   * @return CommonLookupDetail containing the common lookup values.
+   */
+  public Mono<CommonLookupValueDetail> getMaritalStatus(String code) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:133]: final

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 25]
```diff
@@ -0,0 +1,466 @@
...
+  }
+
+  /**
+   * Get a Country Common Value.
+   *
+   * @param code The code of the common lookup values
+   * @return CommonLookupValueDetail containing the common lookup value.
+   */
+  public Mono<CommonLookupValueDetail> getCountry(String code) {
+    return getCountries()
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:166]: final

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 26]
```diff
@@ -0,0 +1,466 @@
...
+    return ebsApiClient.getCommonValues(COMMON_VALUE_CORRESPONDENCE_LANGUAGE);
+  }
+
+  /**
+   * Get a Correspondence Language Common Values.
+   *
+   * @param code The code of the common lookup values
+   * @return CommonLookupValueDetail containing the common lookup value.
+   */
+  public Mono<CommonLookupValueDetail> getCorrespondenceLanguage(String code) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:213]: final

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 27]
```diff
@@ -0,0 +1,466 @@
...
+    return ebsApiClient.getCommonValues(COMMON_VALUE_ETHNIC_ORIGIN);
+  }
+
+  /**
+   * Get an Ethnic Origin Common Values.
+   *
+   * @param code The code of the common lookup values
+   * @return CommonLookupValueDetail containing the common lookup value.
+   */
+  public Mono<CommonLookupValueDetail> getEthnicOrigin(String code) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:232]: final

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 28]
```diff
@@ -0,0 +1,466 @@
...
+    return ebsApiClient.getCommonValues(COMMON_VALUE_DISABILITY);
+  }
+
+  /**
+   * Get a Disability Common Values.
+   *
+   * @param code The code of the common lookup values
+   * @return CommonLookupValueDetail containing the common lookup value.
+   */
+  public Mono<CommonLookupValueDetail> getDisability(String code) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:251]: final

**Review by PhilDigitalJustice — CHANGES_REQUESTED** [review]
Few more comments to resolve

File: src/main/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapper.java (hunk 20 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapper.java:318]: No - there's a bug CCLS-1895 to add the mandatory flag to PriorAuthorityDetail. I've pulled it into the sprint.

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 1 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 2 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 3 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 4 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:104]: done

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 5 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 6 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 7 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 9 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 10 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 11 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 12 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 13 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:769]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 14 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java:810]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 15 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java (hunk 16 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/ApplicationService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 29]
```diff
@@ -43,6 +56,16 @@ public Mono<CommonLookupDetail> getApplicationTypes() {
     return ebsApiClient.getCommonValues(COMMON_VALUE_APPLICATION_TYPE);
   }
 
+  /**
+   * Get a single Application Type Common Value.
+   *
+   * @param code - the code for the application type
+   * @return CommonLookupDetail containing the common lookup values.
+   */
+  public Mono<CommonLookupValueDetail> getApplicationType(String code) {
```
[the remark this answers is not in the document]
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 17 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:300]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 18 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 30]
```diff
@@ -258,7 +256,172 @@ public Mono<CommonLookupDetail> getDisabilities() {
...
+  }
+
+  /**
+   * Retrieves court details.
+   * A wildcard match is performed for both courtCode and description to return all Courts
+   * which contain the provided values.
+   *
+   * @return A Mono containing the CommonLookupDetail or an error handler if an error occurs.
+   */
+  public Mono<CommonLookupDetail> getCourts(String courtCode, String description) {
```
[the remark this answers is not in the document]
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 31]
```diff
@@ -258,7 +256,172 @@ public Mono<CommonLookupDetail> getDisabilities() {
...
+  }
+
+  /**
+   * Retrieves court details.
+   * A wildcard match is performed for courtCode to return all Courts
+   * which contain the provided value.
+   *
+   * @return A Mono containing the CommonLookupDetail or an error handler if an error occurs.
+   */
+  public Mono<CommonLookupDetail> getCourts(String courtCode) {
```
[the remark this answers is not in the document]
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 32]
```diff
@@ -258,7 +256,172 @@ public Mono<CommonLookupDetail> getDisabilities() {
...
+      String outcomeResult) {
+    return ebsApiClient.getOutcomeResults(proceedingCode, outcomeResult);
+  }
+
+  /**
+   * Retrieves stage end data based on the supplied proceedingCode and stageEnd value.
+   *
+   * @return A Mono containing the StageEndLookupDetail or an error handler if an error occurs.
+   */
+  public Mono<StageEndLookupDetail> getStageEnds(String proceedingCode,
```
[the remark this answers is not in the document]
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 33]
```diff
@@ -258,7 +256,172 @@ public Mono<CommonLookupDetail> getDisabilities() {
...
+    return this.getPriorAuthorityTypes(null, null);
+  }
+
+  /**
+   * Retrieves prior authority types matching the specified code and valueRequired flag.
+   *
+   * @return A Mono containing the PriorAuthorityLookupDetail
+   *     or an error handler if an error occurs.
+   */
+  public Mono<PriorAuthorityTypeDetails> getPriorAuthorityTypes(
```
[the remark this answers is not in the document]
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:390]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 34]
```diff
@@ -258,7 +256,172 @@ public Mono<CommonLookupDetail> getDisabilities() {
...
+    return this.getAwardTypes(null, null);
+  }
+
+  /**
+   * Retrieves award types matching the specified code and awardType values.
+   *
+   * @return A Mono containing the AwardTypeLookupDetail
+   *     or an error handler if an error occurs.
+   */
+  public Mono<AwardTypeLookupDetail> getAwardTypes(
```
[the remark this answers is not in the document]
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:413]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 35]
```diff
@@ -258,7 +256,172 @@ public Mono<CommonLookupDetail> getDisabilities() {
...
+
+  /**
+   * Get a single common value based on type and code.
+   *
+   * @param type - the value type.
+   * @param code - the value code.
+   * @return a Mono containing the CommonLookupValueDetail
+   *     or an error handler if an error occurs.
+   */
+  public Mono<CommonLookupValueDetail> getCommonValue(String type, String code) {
```
[the remark this answers is not in the document]
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/resources/application-local.yml
Comment (porritta) [reply src/main/resources/application-local.yml:1]: reverted

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 19 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:223]: added

File: src/main/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapper.java (hunk 20 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapper.java:320]: mandatory flag now mapped correctly

File: src/main/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapper.java (hunk 21 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapper.java:490]: recoveryOfAwardTimeRelated no corrected and mapper updated

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 22 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 23 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 24 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 25 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java:166]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java [hunk 36]
```diff
@@ -0,0 +1,466 @@
...
+    return ebsApiClient.getCommonValues(COMMON_VALUE_CORRESPONDENCE_METHOD);
+  }
+
+  /**
+   * Get a Correspondence Method Common Values.
+   *
+   * @param code The code of the common lookup values
+   * @return CommonLookupValueDetail containing the common lookup value.
+   */
+  public Mono<CommonLookupValueDetail> getCorrespondenceMethod(String code) {
```
[the remark this answers is not in the document]
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 26 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 27 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java (hunk 28 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/service/LookupService.java]: done

File: src/test/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapperTest.java [hunk 37]
```diff
@@ -0,0 +1,1043 @@
...
+
+    ReferenceDataItem result = applicationMapper.toReferenceDataItem(
+        Pair.of(priorAuthorityDetail, priorAuthLookup));
+
+    assertNotNull(result);
+    assertEquals(priorAuthorityDetail.getCode(), result.getCode().getId());
+    assertEquals(priorAuthorityDetail.getDescription(), result.getCode().getDisplayValue());
+    assertEquals(priorAuthorityDetail.getDataType(), result.getType());
+    assertEquals(priorAuthorityDetail.getLovCode(), result.getLovLookUp());
+//    assertEquals(priorAuthorityDetail.get(), result.getMandatory());
```
Comment (PhilDigitalJustice) [inline src/test/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapperTest.java:633]: Do you need this?

**Review by PhilDigitalJustice — CHANGES_REQUESTED** [review]
1 additional comment

File: src/test/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapperTest.java (hunk 37 above)
Comment (porritta) [reply src/test/java/uk/gov/laa/ccms/caab/mapper/ApplicationMapperTest.java]: updated

**Review by PhilDigitalJustice — APPROVED** [review]
LGTM

**Pull request merged**